api/src/infrastructure/kubernetes/infrastructure.rs

```diff
@@ -661,47 +685,68 @@ impl TryFrom<Deployment> for ServiceBuilder {
-        };
+        })?;
+
+        if let (Some(labels), Some(annotations)) = (
+            &deployment.metadata.labels,
+            &deployment.metadata.annotations,
+        ) {
+            let service_name = match labels.get(SERVICE_NAME_LABEL) {
+                Some(service_name) => service_name,
+                None => {
+                    return Err(KubernetesInfrastructureError::MissingServiceNameLabel {
+                        deployment_name: deployment_name.clone(),
+                    });
+                }
+            };
 
-        let image = deployment
-            .metadata
-            .annotations
-            .get(IMAGE_LABEL)
-            .map(|image| {
-                Image::from_str(image).expect("Kubernetes API should provide valid image string")
-            })
-            .ok_or_else(|| KubernetesInfrastructureError::MissingImageLabel {
-                deployment_name: deployment.metadata.name.clone(),
-            })?;
+            let image = annotations
+                .get(IMAGE_LABEL)
+                .map(|image| {
+                    Image::from_str(image)
+                        .expect("Kubernetes API should provide valid image string")
+                })
+                .ok_or_else(|| KubernetesInfrastructureError::MissingImageLabel {
+                    deployment_name: deployment_name.clone(),
+                })?;
 
-        let mut config = ServiceConfig::new(service_name.clone(), image);
+            let mut config = ServiceConfig::new(service_name.clone(), image);
 
-        if let Some(replicated_env) = deployment.metadata.annotations.get(REPLICATED_ENV_LABEL) {
-            let env = serde_json::from_str::<Environment>(replicated_env).map_err(|err| {
-                KubernetesInfrastructureError::UnexpectedError {
-                    internal_message: err.to_string(),
-                }
-            })?;
-            config.set_env(Some(env));
-        }
+            if let Some(replicated_env) = annotations.get(REPLICATED_ENV_LABEL) {
+                let env = serde_json::from_str::<Environment>(replicated_env).map_err(|err| {
+                    KubernetesInfrastructureError::UnexpectedError {
+                        internal_message: err.to_string(),
+                    }
+                })?;
+                config.set_env(Some(env));
+            }
+
+            if let Some(lb) = labels.get(CONTAINER_TYPE_LABEL) {
+                config.set_container_type(lb.parse::<ContainerType>()?);
+            }
 
-        if let Some(lb) = labels.get(CONTAINER_TYPE_LABEL) {
-            config.set_container_type(lb.parse::<ContainerType>()?);
+            Ok(config)
+        } else {
+            Err(KubernetesInfrastructureError::UnexpectedError {
+                internal_message: String::from("Missing deployment labels, annotations, or both"),
+            })
         }
+    }
+}
 
-        Ok(config)
+impl From<serde_json::Error> for KubernetesInfrastructureError {
```

Comment:
I think this could be removed if work internally with `k8s_openapi` values instead of strings